### ISSUE TITLE
Default answer can be overridden by doReturn(..), but not when(..)

### ISSUE THREAD

**Issue by runeflobakk** [mockito/mockito#1381]
I just discovered that mocks with default `Answer`s can not have their behavior overridden using stubbing with `when(..)`, only with `doReturn(..)`. The javadoc for [Mockito.mock(Class, Answer)](https://static.javadoc.io/org.mockito/mockito-core/2.18.3/org/mockito/Mockito.html#mock-java.lang.Class-org.mockito.stubbing.Answer-) says that "It is the default answer so it will be used only when you don't stub the method call.", but this seems to only be the case if I stub method calls using `doReturn` (or any `doXyz`-method I suppose), but not when stubbing with `when(..)`.

The following code demonstrates this:
```java
abstract class MyInterface {
    abstract byte[] getSomeData();
}
MyInterface mock = mock(MyInterface.class, i -> {
    throw new UnsupportedOperationException("method: " + i.getMethod());
});
when(mock.getSomeData()).thenReturn(new byte[0]);
//doReturn(new byte[0]).when(mock).getSomeData();

mock.getSomeData();
```
(Can be pasted into any test method. I used an abstract class to have the code self-contained for a method scope, but the behavior is the same for an interface.)

The invocation of `mock.getSomeData()` will invoke the default answer and throw the `UnsupportedOperationException`. If you comment out the `when`-stubbing, and uncomment the `doReturn`-stubbing, the `mock.getSomeData()` yields the expected stubbed return value.

I have tested this with v2.10.0 and v2.18.3, and I observe the same behavior.

**Comment by runeflobakk** [comment]
This is embarassing, but I of course see what the problem is. The exception is of course thrown when doing the stubbing with `when(..)`, not at the last `mock.getSomeData()`.

I am closing this. Sorry for the noise.

**Comment by TimvdLippe** [comment]
Not an embarrassment at all! We have had multiple of these reports and we are looking for alternatives to resolve this situation. Good luck mocking :)

**Comment by techpavan** [comment]
@TimvdLippe  Any luck on the alternatives so far? That would probably help in making things simple.

**Comment by TimvdLippe** [comment]
No lick yet, sorry :(

**Comment by techpavan** [comment]
Figured out that `when` can be made to work by explicitly returning `Answers.RETURNS_DEFAULTS.answer(invocation)` for all the cases other than our pointed target. More on how I landed here and the solution - https://stackoverflow.com/q/55379745/3940047

**Comment by paulpaulych** [comment]
@TimvdLippe, is there any solution currently found?

**Comment by TimvdLippe** [comment]
There is no solution at the moment sorry and that's unlikely to change for the time being, given fundamental limitations in the Java language. Using `doReturn` is a valid solution for now.

**Comment by laymain** [comment]
Being able to set te default answer **after** stubbing could be a solution, what do you think?

**Comment by feliksik** [comment]
I think I'm running into the same issue. 

I think setting the default after the mocking is not optimal for my usecase, as I want to have it set when creating the mock, so you don't forget to set the default. 

I want a mock that fails hard if there is no explicit stub set. Returning null has never made sense to me. So I want to create every mock like `MyClass mocked = safeMock(MyClass.class)`, like below. 

```
public class HandyMocking {
    /**
     * If you don't stub a response, mockito just returns null, which causes an unexpected NPE elsewhere that makes your
     * System Under Test look broken; but the problem is in your test setup.
     *
     * This method gives you a bit more explicit feedback.
     *
     * But note, it looks like the popular when(myMock.callMethod()).thenReturn(answer) does not work;
     * rather, use doReturn(answer).when(myMock).callMethod().
     */
    public static <T> T safeMock(Class<T> type) {
        return mock(
                type,
                (invocation) -> {
                    throw new RuntimeException("It looks like you forgot to provide a (matching) stub for a method: " + invocation.getMethod());
                }
        );
    }

}
``` 


Somewhere else here I found a question regarding this  that I should use the defaultAnswer, but now I discover it only works the the doReturn().when() syntax, and not with the when().doReturn().

**Comment by renannprado** [comment]
I've just faced the same issue as @feliksik. Default answer behavior of `doReturn().when()` is not the same as `when().doReturn()`.

Our use case here is the following: by default the mocks return null (same as @feliksik it doesn't make sense to me because that's a foot gun). Because we're using kotlin and we rely on compile time check for nullability, we decided that if the method is not mocked, that we should throw an exception instead.

After tinkering a bit, I've found the above answer, changed the syntax and it finally worked.

But from my point of view, this is an unexpected behavior.
